### PR TITLE
Rate limiter

### PR DESCRIPTION
**Pull request by dylanlott**
Allows in-line rate limiting of any route, with custom per-route settings and configs

### CODE REVIEW

**Comment by braydonf** [comment]
Probably need to update `travis.yml` with https://docs.travis-ci.com/user/database-setup/#Redis

**Comment by braydonf** [comment]
Ran through a quick benchmark between redis, memcached and mongodb, and here are the results:
```
redis#set x 33,623 ops/sec ±1.86% (75 runs sampled)
memcached#set x 13,723 ops/sec ±1.46% (76 runs sampled)
mongodb#set x 6,887 ops/sec ±7.92% (63 runs sampled)
redis#get x 37,549 ops/sec ±1.73% (81 runs sampled)
memcached#get x 14,970 ops/sec ±1.64% (79 runs sampled)
mongodb#get x 6,112 ops/sec ±12.57% (60 runs sampled)
Fastest is redis#get
```
Here is the script that was used: https://gist.github.com/braydonf/ff1f2b3352ae80a86a41a3644a111670

File: README.md [hunk 1]
```diff
@@ -16,3 +15,47 @@ var app = require('express')();
...
+  total: 150,
+  expire: 1000 * 60 // 150 requests a minute allowed
+  }), function(req, res) {
+  res.status(200).send('hello');
+});
+```
+
+# Testing
+
+Redis must be running for the tests to work. The easiest way to get an instance running is via Docker.
```
Comment (braydonf) [inline README.md]: Also `apt-get install redis-server`

Comment (pgerbes1) [inline README.md]: Nice, Redis is awesome and ridiculously fast. Happy to see us using it.

**Comment by braydonf** [comment]
express needs to be added to devDependencies: `npm install express --save-dev`

File: lib/rate-limiter.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        };
+      opts.keyFormatter = typeof opts.keyFormatter === 'function'
+        ? opts.keyFormatter
+        : function(parts) {
+          return "rateLimit:" + parts.join(':');
+        };
+      var key = opts.keyFormatter(opts.lookup(req));
+      db.get(key, function(err, limit) {
+        if (err && opts.ignoreErrors) {
+          return next();
```
Comment (braydonf) [inline lib/rate-limiter.js]: We should be able to log warnings here

File: lib/rate-limiter.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          };
+
+        if (now > limit.reset) {
+          limit.reset = now + opts.expire;
+          limit.remaining = opts.total;
+        }
+
+        // do not allow negative remaining
+        limit.remaining = Math.max(Number(limit.remaining) - 1, -1);
+        db.set(key, JSON.stringify(limit), 'PX', opts.expire, function() {
```
Comment (braydonf) [inline lib/rate-limiter.js]: Can there be an error from this?

File: lib/rate-limiter.js (hunk 2 above)
Comment (dylanlott) [inline lib/rate-limiter.js]: I can add some logic to check if it's below a configurable threshold or something.

Comment (braydonf) [inline lib/rate-limiter.js]: For reference, we have a logger for the error middleware by passing the logger in as an option: https://github.com/Storj/service-middleware/blob/master/lib/error-handler.js#L8

**Comment by braydonf** [comment]
Going to go ahead and merge this in, and we can work on it from there.

**Pull request merged**